render/render_test.go

```diff
@@ -18,73 +18,154 @@ const (
+     - "8080:80"
+`)
+}
+
+func TestSubstituteNamedParams(t *testing.T) {
+	composeFile := `
+version: "3.6"
+services:
+  back:
+    ports:
+     - "$back.port:90"
+`
+	parameters := map[string]string{
+		"back.port": "9000",
 	}
-	_, err := render("foo.dockerapp", configFiles, finalEnv, nil)
-	assert.Check(t, err != nil)
-	assert.Check(t, is.ErrorContains(err, "required variable"))
+	s, err := substituteParams(parameters, composeFile)
+	assert.NilError(t, err)
+	assert.Equal(t, s, `
+version: "3.6"
+services:
+  back:
+    ports:
+     - "9000:90"
+`)
 }
 
-func TestRender(t *testing.T) {
-	configFiles := []composetypes.ConfigFile{
-		{
-			Config: map[string]interface{}{
-				"version": "3",
-				"services": map[string]interface{}{
-					"foo": map[string]interface{}{
-						"image":   "busybox:${version}",
-						"command": []interface{}{"-text", "${foo.bar}"},
-					},
-				},
-			},
-		},
+func TestRenderFailOnDefaultParamValueInCompose(t *testing.T) {
+	metadata := strings.NewReader(validMeta)
+	composeFile := strings.NewReader(`
+version: "3.6"
+services:
+  front:
+    ports:
+     - "${front.port:-9090}:80"
+`)
+	app := &types.App{Path: "my-app"}
+	assert.NilError(t, types.Metadata(metadata)(app))
+	assert.NilError(t, types.WithComposes(composeFile)(app))
+	userParameters := map[string]string{
+		"front.port": "4242",
 	}
-	finalEnv := map[string]string{
-		"version": "latest",
-		"foo.bar": "baz",
+	_, err := Render(app, userParameters, nil)
+	assert.ErrorContains(t, err, "Parameters must not have default values set in compose file. Invalid parameter: ${front.port:-9090}.")
+}
+
+func TestSubstituteMixedParams(t *testing.T) {
+	composeFile := `
+version: "3.6"
+services:
+  front:
+    ports:
+     - "${front.port}:80"
+    deploy:
+      replicas: ${front-deploy.replicas}
+  back:
+    ports:
+     - "$back.port:90"
+`
+	parameters := map[string]string{
+		"front.port":            "8080",
+		"back.port":             "9000",
+		"front-deploy.replicas": "3",
 	}
-	c, err := render("foo.dockerapp", configFiles, finalEnv, nil)
+	s, err := substituteParams(parameters, composeFile)
 	assert.NilError(t, err)
-	assert.Check(t, is.Len(c.Services, 1))
-	assert.Check(t, is.Equal(c.Services[0].Image, "busybox:latest"))
-	assert.Check(t, is.DeepEqual([]string(c.Services[0].Command), []string{"-text", "baz"}))
+	assert.Equal(t, s, `
+version: "3.6"
+services:
+  front:
+    ports:
+     - "8080:80"
+    deploy:
+      replicas: 3
+  back:
+    ports:
+     - "9000:90"
+`)
+}
+
+func TestSubstituteSkipDoubleDollarCase(t *testing.T) {
+	composeFile := `
+	version: "3.7"
+	services:
+	  front:
+		command: monitor --source ${app.name}-$app.version $$dollar
+`
+	parameters := map[string]string{
+		"app.name":    "test",
+		"app.version": "version",
+	}
+	s, err := substituteParams(parameters, composeFile)
+	assert.NilError(t, err)
+	assert.Equal(t, s, `
+	version: "3.7"
+	services:
+	  front:
+		command: monitor --source test-version $$dollar
```

Comment:
This line makes me think about what if we would like to print `$dollar` instead of `$$dollar` how could I write it?

Reply:
Maybe setting a parameter with the value "$dollar" and in compose use the parameter name instead of "$$dollar".